File: src/services/staff.ts
import { API_ROUTES } from "@/shared/constants/api";
import { MAX_PAGE_SIZE } from "@/shared/constants/config";
import { apiRequest, queryString, type PaginatedResponse } from "@/services/api";

export interface StaffLedgerEntry {
  id: string;
  userId: string;
  userName?: string | null;
  branchId: string;
  entryType: string;
  moneyTransactionId?: string | null;
  debit: number;
  credit: number;
  balanceAfter: number;
  notes?: string | null;
  occurredAt: string;
  createdAt: string;
}

export interface StaffLedgerResponse {
  userId: string;
  totalPaid: number;
  paidThisMonth: number;
  entries: StaffLedgerEntry[];
}

export interface StaffPayoutResponse {
  id: string;
  userId: string;
  amount: number;
  payoutType: string;
  paymentMethod: string;
  moneyTransactionId: string;
  balanceAfter: number;
  occurredAt: string;
}

export function listStaffLedgers(
  params: {
    page?: number;
    perPage?: number;
    search?: string;
    user?: string;
    entryType?: string;
    notes?: string;
    debit?: number;
    occurredFrom?: string;
    occurredTo?: string;
  } = {},
  signal?: AbortSignal,
) {
  return apiRequest<PaginatedResponse<StaffLedgerEntry>>(
    `${API_ROUTES.staffLedgers}${queryString(params)}`,
    { signal },
  );
}

export async function listAllStaffLedgers(
  params: Omit<Parameters<typeof listStaffLedgers>[0], "page" | "perPage"> = {},
  signal?: AbortSignal,
) {
  const rows: StaffLedgerEntry[] = [];
  let page = 1;
  for (;;) {
    const response = await listStaffLedgers({ ...params, page, perPage: MAX_PAGE_SIZE }, signal);
    rows.push(...response.data);
    if (!response.meta.hasNextPage) return rows;
    page += 1;
  }
}

export function getStaffLedger(userId: string, signal?: AbortSignal) {
  return apiRequest<StaffLedgerResponse>(API_ROUTES.userLedger(userId), { signal });
}

export function recordStaffPayout(
  userId: string,
  payload: {
    amount: number;
    payoutType: "SALARY" | "COMMISSION";
    paymentMethod?: string;
    notes?: string | null;
  },
) {
  return apiRequest<StaffPayoutResponse>(API_ROUTES.userPayouts(userId), {
    method: "POST",
    body: JSON.stringify({
      amount: payload.amount,
      payoutType: payload.payoutType,
      paymentMethod: payload.paymentMethod ?? "CASH",
      notes: payload.notes,
    }),
  });
}

File: src-tauri/src/backend/dto/staff.rs
use rust_decimal::Decimal;
use serde::{Deserialize, Serialize};
use validator::Validate;

use super::PageQuery;

#[derive(Clone, Debug, Default, Deserialize)]
#[serde(default, rename_all = "camelCase")]
pub struct StaffLedgerListQuery {
    #[serde(flatten)]
    pub page: PageQuery,
    pub user: Option<String>,
    pub entry_type: Option<String>,
    pub debit: Option<f64>,
    pub credit: Option<f64>,
    pub occurred_from: Option<String>,
    pub occurred_to: Option<String>,
}

#[derive(Clone, Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct StaffLedgerEntryResponse {
    pub id: String,
    pub user_id: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub user_name: Option<String>,
    pub branch_id: String,
    pub entry_type: String,
    pub money_transaction_id: Option<String>,
    pub debit: f64,
    pub credit: f64,
    pub balance_after: f64,
    pub notes: Option<String>,
    pub occurred_at: String,
    pub created_at: String,
}

#[derive(Clone, Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct StaffLedgerResponse {
    pub user_id: String,
    pub total_paid: f64,
    pub paid_this_month: f64,
    pub entries: Vec<StaffLedgerEntryResponse>,
}

#[derive(Clone, Debug, Deserialize, Validate)]
#[serde(rename_all = "camelCase")]
pub struct StaffPayoutRequest {
    #[validate(custom(function = "positive_decimal"))]
    pub amount: Decimal,
    #[validate(length(min = 1, max = 32))]
    pub payout_type: String,
    #[serde(default = "default_payment_method")]
    pub payment_method: String,
    #[validate(length(max = 1_000))]
    #[serde(default)]
    pub notes: Option<String>,
}

#[derive(Clone, Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct StaffPayoutResponse {
    pub id: String,
    pub user_id: String,
    pub amount: f64,
    pub payout_type: String,
    pub payment_method: String,
    pub money_transaction_id: String,
    pub balance_after: f64,
    pub occurred_at: String,
}

fn default_payment_method() -> String {
    "CASH".to_owned()
}

fn positive_decimal(value: &Decimal) -> Result<(), validator::ValidationError> {
    if *value > Decimal::ZERO {
        Ok(())
    } else {
        Err(validator::ValidationError::new("positive_amount"))
    }
}
